### PR TITLE
[amp] Add __all__ to torch.amp

### LINKED ISSUE #181654: __all__ not implemented for torch.amp
### 🐛 Describe the bug

The `torch.amp` module does not have an `__all__` variable defined, causing type checkers like Pylance to report problems like `"GradScaler" is not exported from module "torch.amp"`. Other modules (i.e. `torch.cpu.amp` and `torch.cuda.amp`) do have this defined, but seeing as `torch.amp` is recommended over these two, it should also be implemented for them.

[#179715](https://github.com/pytorch/pytorch/pull/179715) implements the fix for this issue, but was closed by the author before being reviewed by any maintainers.

### Versions

Collecting environment information...
PyTorch version: 2.11.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.2 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.3.2)
CMake version: version 4.2.3
Libc version: N/A

Python version: 3.14.0 (main, Oct 31 2025, 23:20:55) [Clang 21.1.4 ] (64-bit runtime)
Python platform: macOS-26.2-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M4

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

cc @lolpack @maggiemoss @ndmitchell @kinto0 @samwgoldman @mcarilli @ptrblck @leslie-fang-intel @jgong5

### PR DESCRIPTION
**Pull request by Priyanshu-byte-coder**
## Summary

Adds an explicit `__all__` to `torch/amp/__init__.py`. Without it, static type checkers (Pylance, Pyright) flag exports like `GradScaler` and `autocast` as unconfirmed re-exports.

Follows the same pattern as `torch/cuda/amp/__init__.py`. Private names `_enter_autocast` and `_exit_autocast` are intentionally excluded from `__all__`.

Fixes #181654

## Test plan
- [ ] `python -c "import torch.amp; print(torch.amp.__all__)"` outputs the expected list
- [ ] No regressions in existing amp tests

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

**Review by malfet — APPROVED** [review]

**Comment by Priyanshu-byte-coder** [comment]
@pytorchbot label "topic: not user facing"

@pytorch/pytorch-dev-infra could you approve the CI workflows for this PR? One file changed (`torch/amp/__init__.py`, +9 lines), no behavior impact — just adds `__all__` to fix static type checker false positives.

**Review by cyyever — APPROVED** [review]
